### ISSUE TITLE
Total energy of G0W0 calculation

### ISSUE THREAD

**Issue by ZhishuoHuang** [pyscf/pyscf#1149]
Dear all,

I am using PySCF to do G0W0 calculations.
I ran the example provided on PySCF forum. 
But even though I set the ‘mol.verbose = 9’, there are only quasiparticle orbital energies given. 
I wonder if it is possible to output the total energy and if possible, how to do that?

I post the input file below:
######################################
#!/usr/bin/env python

'''
A simple example to run a GW calculation
'''

from pyscf import gto, dft, gw
mol = gto.M(
            atom = 'H 0 0 0; F 0 0 1.1',
                basis = 'ccpvdz')
mol.output = '/home/scratch/zhishuo/PySCF/TEST_examples/GW'
mol.verbose = 9
mf = dft.RKS(mol)
mf.xc = 'pbe'
mf.kernel()

nocc = mol.nelectron//2

# By default, GW is done with analytic continuation
gw = gw.GW(mf)
# same as gw = gw.GW(mf, freq_int='ac')
gw.kernel()
print(gw.mo_energy)
######################################

with the output file shown as following:
######################################

******** <class 'pyscf.dft.rks.RKS'> ********
method = RKS-RHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 50
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /home/scratch/zhishuo/PySCF/TEST_examples/GW/tmpbawer1v3
max_memory 4000 MB (current use 75 MB)
XC library pyscf.dft.libxc version 5.1.7
    S. Lehtola, C. Steigemann, M. J. Oliveira, and M. A. Marques, SoftwareX 7, 1 (2018)
XC functionals = pbe
    J. P. Perdew, K. Burke, and M. Ernzerhof, Phys. Rev. Lett. 77, 3865 (1996)
    J. P. Perdew, K. Burke, and M. Ernzerhof, Phys. Rev. Lett. 78, 1396 (1997)
    J. P. Perdew, K. Burke, and M. Ernzerhof, Phys. Rev. Lett. 77, 3865 (1996)
    J. P. Perdew, K. Burke, and M. Ernzerhof, Phys. Rev. Lett. 78, 1396 (1997)
radial grids:
    Treutler-Ahlrichs [JCP 102, 346 (1995); DOI:10.1063/1.469408] (M4) radial grids

becke partition: Becke, JCP 88, 2547 (1988); DOI:10.1063/1.454033
pruning grids: <function nwchem_prune at 0x7f4096f4e040>
grids dens level: 3
symmetrized grids: False
atomic radii adjust function: <function treutler_atomic_radii_adjust at 0x7f4096f3ff70>
atomic_radii : [3.77945036 0.66140414 2.64561657 2.74010288 1.98421243 1.60626721
 1.32280829 1.22832198 1.13383567 0.94486306 2.83458919 3.40150702
 2.83458919 2.36215766 2.07869874 1.88972612 1.88972612 1.88972612
 3.40150702 4.15739747 3.40150702 3.0235618  2.64561657 2.55113027
 2.64561657 2.64561657 2.64561657 2.55113027 2.55113027 2.55113027
 2.55113027 2.45664396 2.36215766 2.17318504 2.17318504 2.17318504
 3.59047964 4.44085639 3.77945225 3.40150702 2.92907549 2.74010288
 2.74010288 2.55113027 2.45664396 2.55113027 2.64561657 3.0235618
 2.92907549 2.92907549 2.74010288 2.74010288 2.64561657 2.64561657
 3.96842486 4.91328792 4.06291117 3.68496594 3.49599333 3.49599333
 3.49599333 3.49599333 3.49599333 3.49599333 3.40150702 3.30702072
 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072
 2.92907549 2.74010288 2.55113027 2.55113027 2.45664396 2.55113027
 2.55113027 2.55113027 2.83458919 3.59047964 3.40150702 3.0235618
 3.59047964 2.74010288 3.96842486 3.40150702 4.06291117 3.68496594
 3.40150702 3.40150702 3.30702072 3.30702072 3.30702072 3.30702072
 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072
 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072
 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072
 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072
 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072
 3.30702072 3.30702072 3.30702072 3.30702072 3.30702072]
small_rho_cutoff = 1e-07
Set gradient conv threshold to 3.16228e-05
Nelec from initial guess = 9.977410093935312
atom H rad-grids = 50, ang-grids = [ 50  50  50  50  50  50  50  50  50  50  50  50  50  50  86  86  86  86
 266 266 266 266 302 302 302 302 302 302 302 302 302 302 302 302 302 266
 266 266 266 266 266 266 266 266 266 266 266 266 266 266]
atom F rad-grids = 75, ang-grids = [ 50  50  50  50  50  50  50  50  50  50  50  50  50  50  50  50  50  50
  50  50  50  86  86  86  86  86  86  86  86  86 266 266 266 266 266 266
 302 302 302 302 302 302 302 302 302 302 302 302 302 302 302 302 302 302
 266 266 266 266 266 266 266 266 266 266 266 266 266 266 266 266 266 266
 266 266 266]
tot grids = 24466
Drop grids 2965
    CPU time for setting up grids      0.02 sec, wall time      0.02 sec
nelec by numeric integration = 9.977410058506097
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -147.79366184144808  Ecoul = 53.81545980555761  Exc = -10.56315078337966
init E= -100.211721093561
cond(S) = 36.03560971202316
    CPU time for initialize scf      0.09 sec, wall time      0.10 sec
  HOMO = -0.433574392955141  LUMO = -0.0442325133887142
  mo_energy =
[-24.42316777  -1.14925327  -0.5181955   -0.43357439  -0.43357439
  -0.04423251   0.43679971   0.94825256   0.94825256   1.04310565
   1.30276609   1.30276609   1.58236173   1.82722031   3.35715328
   3.35715413   3.48370399   3.48370399   4.12568286]
nelec by numeric integration = 9.999999954153106
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -151.0872629864317  Ecoul = 57.46381765256835  Exc = -10.95113537035863
cycle= 1 E= -100.244948978513  delta_E= -0.0332  |g|= 0.628  |ddm|= 0.912
    CPU time for cycle= 1      0.04 sec, wall time      0.04 sec
diis-norm(errvec)=1.02923
diis-c [-1.05932303  1.        ]
  HOMO = -0.123004851649509  LUMO = 0.0172140447828145
  mo_energy =
[-2.38123704e+01 -8.61303203e-01 -3.14577195e-01 -1.23004852e-01
 -1.23004852e-01  1.72140448e-02  4.33449184e-01  1.09442013e+00
  1.09442013e+00  1.27256228e+00  1.37012124e+00  1.37012124e+00
  1.69046388e+00  1.96509937e+00  3.68634249e+00  3.68634349e+00
  3.78246750e+00  3.78246750e+00  4.38773121e+00]
nelec by numeric integration = 9.999999942292893
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -144.99518711059403  Ecoul = 50.830645460069775  Exc = -10.20529865697312
cycle= 2 E= -100.040208581788  delta_E= 0.205  |g|= 1.25  |ddm|= 1.05
    CPU time for cycle= 2      0.03 sec, wall time      0.03 sec
diis-norm(errvec)=2.02419
diis-c [-0.03198082  0.66546961  0.33453039]
  HOMO = -0.296437558761834  LUMO = -0.0318248464538084
  mo_energy =
[-24.23667377  -1.02673171  -0.42729085  -0.29643756  -0.29643756
  -0.03182485   0.43198499   1.01987653   1.01987653   1.14175934
   1.31753693   1.31753693   1.62424315   1.87830502   3.507597
   3.50759796   3.60746965   3.60746965   4.22947123]
nelec by numeric integration = 9.999999949793388
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -149.21611744191034  Ecoul = 55.26765518445286  Exc = -10.698087527570502
cycle= 3 E= -100.316918059319  delta_E= -0.277  |g|= 0.0213  |ddm|= 0.723
    CPU time for cycle= 3      0.03 sec, wall time      0.03 sec
diis-norm(errvec)=0.0437793
diis-c [-3.81030875e-04  3.78042053e-02  3.57987417e-02  9.26397053e-01]
  HOMO = -0.292325571653969  LUMO = -0.0332430466973048
  mo_energy =
[-24.23517775  -1.02185256  -0.42455774  -0.29232557  -0.29232557
  -0.03324305   0.42891064   1.02131079   1.02131079   1.14461189
   1.31701752   1.31701752   1.62592025   1.8799779    3.51228341
   3.51228436   3.61268811   3.61268811   4.23368417]
nelec by numeric integration = 9.999999949142088
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -149.14948106689314  Ecoul = 55.19357036944616  Exc = -10.690773726178836
cycle= 4 E= -100.317052697917  delta_E= -0.000135  |g|= 0.00284  |ddm|= 0.0185
    CPU time for cycle= 4      0.03 sec, wall time      0.03 sec
diis-norm(errvec)=0.00404271
diis-c [-4.86602792e-06  3.80053251e-04  3.08298350e-03  1.06646958e-01
  8.89890005e-01]
  HOMO = -0.292116387532438  LUMO = -0.0333014864608095
  mo_energy =
[-24.23349534  -1.02179203  -0.42426063  -0.29211639  -0.29211639
  -0.03330149   0.42895946   1.02156708   1.02156708   1.14475524
   1.31706167   1.31706167   1.62592571   1.88001411   3.51240356
   3.51240451   3.61272772   3.61272772   4.23375717]
nelec by numeric integration = 9.999999949190618
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -149.14500123850584  Ecoul = 55.18834477489713  Exc = -10.690029108801216
cycle= 5 E= -100.317053846701  delta_E= -1.15e-06  |g|= 0.000271  |ddm|= 0.00173
    CPU time for cycle= 5      0.03 sec, wall time      0.03 sec
diis-norm(errvec)=0.000520753
diis-c [-2.58571251e-08 -1.09777162e-04  7.51218634e-06 -6.75273608e-03
  8.64788948e-02  9.20376106e-01]
  HOMO = -0.292130792516525  LUMO = -0.0333531699225601
  mo_energy =
[-24.23351507  -1.0218054   -0.42432539  -0.29213079  -0.29213079
  -0.03335317   0.42888589   1.02153383   1.02153383   1.14471416
   1.31700999   1.31700999   1.62588855   1.87995637   3.51239195
   3.51239291   3.61269377   3.61269377   4.23371024]
nelec by numeric integration = 9.999999949185455
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -149.14466195985864  Ecoul = 55.18797496026121  Exc = -10.689998587271836
cycle= 6 E= -100.31705386116  delta_E= -1.45e-08  |g|= 2.91e-05  |ddm|= 0.000155
    CPU time for cycle= 6      0.03 sec, wall time      0.03 sec
diis-norm(errvec)=5.26053e-05
diis-c [-6.89065592e-10 -3.73588694e-05 -9.46190378e-06 -1.37443837e-03
  7.88163419e-03  1.52532667e-01  8.41006958e-01]
  HOMO = -0.292133728435749  LUMO = -0.0333585859092196
  mo_energy =
[-24.23353017  -1.0218092   -0.42432717  -0.29213373  -0.29213373
  -0.03335859   0.42887909   1.0215283    1.0215283    1.1447179
   1.31700436   1.31700436   1.62587931   1.87995254   3.51238982
   3.51239078   3.61269331   3.61269331   4.23370895]
nelec by numeric integration = 9.999999949184097
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -149.14465652319063  Ecoul = 55.18797201653816  Exc = -10.69000108041619
cycle= 7 E= -100.31705386136  delta_E= -1.99e-10  |g|= 8.54e-06  |ddm|= 3.32e-05
    CPU time for cycle= 7      0.03 sec, wall time      0.03 sec
  HOMO = -0.292130748240692  LUMO = -0.033355737045129
  mo_energy =
[-24.23352062  -1.02180575  -0.42432416  -0.29213075  -0.29213075
  -0.03335574   0.42888222   1.02153141   1.02153141   1.1447189
   1.31700698   1.31700698   1.62588325   1.87995494   3.51239239
   3.51239335   3.61269597   3.61269597   4.23371182]
nelec by numeric integration = 9.999999949184428
    CPU time for vxc      0.03 sec, wall time      0.03 sec
E1 = -149.14461413263734  Ecoul = 55.187923145283825  Exc = -10.689994599717537
Extra cycle  E= -100.317053861362  delta_E= -2.4e-12  |g|= 8.47e-06  |ddm|= 9.75e-06
    CPU time for scf_cycle      0.35 sec, wall time      0.35 sec
    CPU time for SCF      0.35 sec, wall time      0.35 sec
converged SCF energy = -100.317053861362
Default auxbasis cc-pvdz-ri is used for F ccpvdz
Default auxbasis cc-pvdz-ri is used for H ccpvdz

******** <class 'pyscf.gw.gw_ac.GWAC'> ********
method = GWAC
GW nocc = 5, nvir = 14
use perturbative linearized QP eqn = False
analytic continuation method = pade
******** <class 'pyscf.df.df.DF'> ********
auxbasis = {'F': 'cc-pvdz-ri', 'H': 'cc-pvdz-ri'}
max_memory = 4000
_cderi_to_save = /home/scratch/zhishuo/PySCF/TEST_examples/GW/tmpgoo607ja
num shells = 24, num cGTOs = 70
size of aux basis 70
    CPU time for 2c2e      0.00 sec, wall time      0.00 sec
shranges = [(0, 8, 190)]
int3c2e [1/1], AO [0:8], nrow = 190
    CPU time for cholesky_eri      0.00 sec, wall time      0.00 sec
nelec by numeric integration = 9.999999949184426
    CPU time for vxc      0.03 sec, wall time      0.03 sec
  GW mo_energy =
[-26.20887482  -1.34556612  -0.63970147  -0.50371319  -0.50371319
   0.13820756   0.61429425   1.29017698   1.2901769    1.38517064
   1.56518126   1.56518087   1.91850741   2.15492359   3.86029428
   3.86028626   4.02315672   4.02315594   4.3233935 ]

WARN: GW QP energies may not be sorted from min to max

    CPU time for GW      0.19 sec, wall time      0.19 sec
[-26.20887482  -1.34556612  -0.63970147  -0.50371319  -0.50371319
   0.13820756   0.61429425   1.29017698   1.2901769    1.38517064
   1.56518126   1.56518087   1.91850741   2.15492359   3.86029428
   3.86028626   4.02315672   4.02315594   4.3233935 ]
######################################

Best regards
Zhishuo Huang

**Comment by tberkel** [comment]
The GW approximation is rarely used to calculate total energies. Most
similar is the RPA total energy, an efficient implementation of which is
not currently in PySCF but is a planned feature.

**Comment by ZhishuoHuang** [comment]
> 

Hello, thank you for your reply. 
There is one approximate way of total energy calculation, of which the total energy is the summation of total energy at RPA level + the Galitskii-Migdal correction. 

So the total energy at RPA level is not currently available in PySCF?

**Comment by tberkel** [comment]
Indeed, given the full GW Green's function (and not just its quasiparticle approximation), one could in principle calculate the total energy by frequency integration, although controlling this integration error is numerically nontrivial.

An *efficient* RPA total energy calculation is not currently available in PySCF. Perhaps @zhutianyu1991, who wrote the latest GW implementation in PySCF, could tell us if/when he plans to contribute RPA total energies.

Two inefficient options are:
1. The plasmon formula for the RPA energy. PySCF does have implementations of RPA and TDA excitation energies, the summed difference of which gives the RPA total energy. But this requires enumerating all excitation energies and so will be O(N^6) cost.

2. The ring CCD version of the RPA energy. PySCF's CCSD codes can be modified to only include ring (or direct ring) diagrams, which also comes with O(N^6) or O(N^5) cost. My group is working a little on this, and has a few hacky codes that we could share. If you're interested in this, you can contact me offline and to discuss further.

**Comment by zhutianyu1991** [comment]
Hi, current GW implementations only compute the quasiparticle energy, and the total energy is not computed. For non-self-consistent GW (G0W0), there're a few definitions of total energy (such as Galitskii-Migdal or Klein) that give different values, but I don't know if there's consensus on which one is the best. 
Instead, I think RPA energy is what many literatures reported as "G0W0 energy". I have an O(N^4) implementation of RPA total energy (which re-uses many routines in GW-AC code) in my local repo, and should probably make a pull request to merge this code into PySCF's main branch now. In the meantime, if you want to use this code sooner, I'll email you about it.

**Comment by ZhishuoHuang** [comment]
> Indeed, given the full GW Green's function (and not just its quasiparticle approximation), one could in principle calculate the total energy by frequency integration, although controlling this integration error is numerically nontrivial.
> 
> An _efficient_ RPA total energy calculation is not currently available in PySCF. Perhaps @zhutianyu1991, who wrote the latest GW implementation in PySCF, could tell us if/when he plans to contribute RPA total energies.
> 
> Two inefficient options are:
> 
> 1. The plasmon formula for the RPA energy. PySCF does have implementations of RPA and TDA excitation energies, the summed difference of which gives the RPA total energy. But this requires enumerating all excitation energies and so will be O(N^6) cost.
> 2. The ring CCD version of the RPA energy. PySCF's CCSD codes can be modified to only include ring (or direct ring) diagrams, which also comes with O(N^6) or O(N^5) cost. My group is working a little on this, and has a few hacky codes that we could share. If you're interested in this, you can contact me offline and to discuss further.

I see. 
Thank you for your detailed explanation.

**Comment by ZhishuoHuang** [comment]
> Hi, current GW implementations only compute the quasiparticle energy, and the total energy is not computed. For non-self-consistent GW (G0W0), there're a few definitions of total energy (such as Galitskii-Migdal or Klein) that give different values, but I don't know if there's consensus on which one is the best. Instead, I think RPA energy is what many literatures reported as "G0W0 energy". I have an O(N^4) implementation of RPA total energy (which re-uses many routines in GW-AC code) in my local repo, and should probably make a pull request to merge this code into PySCF's main branch now. In the meantime, if you want to use this code sooner, I'll email you about it.

Hi, could you please email me the code, I want to use it sooner. 
My email address is: zhishuohuang@gmail.com
I will appreciate that.

**Comment by verena-neufeld** [comment]
Hi @ZhishuoHuang , has your question been answered?

**Comment by ZhishuoHuang** [comment]
> Hi @ZhishuoHuang , has your question been answered?
Yes, to some extent I know that currently, total energy is only available at RPA level.

**Comment by hungpham2017** [comment]
> Indeed, given the full GW Green's function (and not just its quasiparticle approximation), one could in principle calculate the total energy by frequency integration, although controlling this integration error is numerically nontrivial.
> 
> An _efficient_ RPA total energy calculation is not currently available in PySCF. Perhaps @zhutianyu1991, who wrote the latest GW implementation in PySCF, could tell us if/when he plans to contribute RPA total energies.
> 
> Two inefficient options are:
> 
> 1. The plasmon formula for the RPA energy. PySCF does have implementations of RPA and TDA excitation energies, the summed difference of which gives the RPA total energy. But this requires enumerating all excitation energies and so will be O(N^6) cost.
> 2. The ring CCD version of the RPA energy. PySCF's CCSD codes can be modified to only include ring (or direct ring) diagrams, which also comes with O(N^6) or O(N^5) cost. My group is working a little on this, and has a few hacky codes that we could share. If you're interested in this, you can contact me offline and to discuss further.

@tberkel: I wonder if the ring CCD is available somewhere in PySCF or it is still an internal code?

**Comment by tberkel** [comment]
No, it's not in PySCF. I don't know if it's of sufficiently broad interest that it needs to be in PySCF. But it's really only a few lines to modify the `update_amps` function. All the t1 stuff can be deleted, and only a few contractions need to be kept for ring CCD. If I have time, maybe I'll write something and place it in the `examples/cc` folder.

**Comment by hungpham2017** [comment]
An example in pyscf/examples/cc would be helpful. I am interested in using it. I'll email you offline for further discussion then. Thanks for the prompt response.

**Comment by changsuc** [comment]
Hi friends, I wrote a demo code by modifying the `update_amps` and `energy` in `pyscf/cc/rccsd_slow.py` according to the discussion above. I put the file here, maybe someone will find it helpful in future. 

```
'''
Restricted direct ring-CCD

Ref: Scuseria et al., J. Chem. Phys. 129, 231101 (2008)
Ref: Masios et al., Phys. Rev. Lett. 131, 186401 (2023)

Original Source:
Modified from pyscf/cc/rccsd_slow.py

Contact: 
caochangsu@gmail.com
'''

from functools import reduce
import numpy as np
from pyscf.cc import ccsd
from pyscf import lib
from pyscf.ao2mo import _ao2mo

einsum = lib.einsum

def update_amps(cc, t1, t2, eris):
    # Ref: Masios et al., Phys. Rev. Lett. 131, 186401 (2023) Eq.(3)
    nocc, nvir = t1.shape
    t1[:] = 0.0

    eris_ovvo = np.asarray(eris.ovvo)
    eris_ovov = np.asarray(eris.ovov)

    t2new = 2*einsum('iack, kjcb->iajb', eris_ovvo, t2)
    t2new += 2*einsum('ikac, jbck->iajb', t2, eris_ovvo)
    tmp = einsum('ikac, kcld->iald', t2, eris_ovov)
    t2new += 4*einsum('iald, ljdb->iajb', tmp, t2)
    t2new = t2new.transpose(0,2,1,3)

    t2new += np.asarray(eris.ovov).conj().transpose(0,2,1,3).copy()

    mo_e = eris.fock.diagonal().real
    eia = mo_e[:nocc,None] - mo_e[None,nocc:]
    eijab = lib.direct_sum('ia,jb->ijab',eia,eia)
    t2new /= eijab

    return t1, t2new

def energy(cc, t1, t2, eris):
    # Ref: Scuseria et al., J. Chem. Phys. 129, 231101 (2008) Eq.(11)
    nocc, nvir = t1.shape
    fock = eris.fock
    eris_ovov = np.asarray(eris.ovov)
    e = 2*einsum('ijab,iajb', t2, eris_ovov)
    return e.real

def init_amps(cc, eris):
    nocc = cc.nocc
    mo_e = eris.fock.diagonal().real
    eia = mo_e[:nocc,None] - mo_e[None,nocc:]
    eijab = lib.direct_sum('ia,jb->ijab', eia, eia)
    t1 = np.zeros_like(eris.fock[:nocc,nocc:])

    eris_ovov = np.asarray(eris.ovov)
    t2 = eris_ovov.transpose(0,2,1,3).conj() / eijab
    cc.emp2  = 2*einsum('ijab,iajb', t2, eris_ovov)
    cc.emp2 -=   einsum('ijab,ibja', t2, eris_ovov)
    lib.logger.info(cc, 'Init t2, MP2 energy = %.15g', cc.emp2)
    return cc.emp2, t1, t2

def ao2mo(cc, mo_coeff=None):
    return _ChemistsERIs(cc, mo_coeff)

class _ChemistsERIs:
    def __init__(self, cc, mo_coeff=None, method='incore'):
        if mo_coeff is None:
            self.mo_coeff = mo_coeff = ccsd._mo_without_core(cc, cc.mo_coeff)
        else:
            self.mo_coeff = mo_coeff = ccsd._mo_without_core(cc, mo_coeff)
        dm = cc._scf.make_rdm1(cc.mo_coeff, cc.mo_occ)
        fockao = cc._scf.get_hcore() + cc._scf.get_veff(cc.mol, dm)
        self.fock = reduce(np.dot, (mo_coeff.T, fockao, mo_coeff))

        nocc = cc.nocc
        nmo = cc.nmo
        nvir = nmo - nocc
        # eri1 = ao2mo.incore.full(cc._scf._eri, mo_coeff)
        # eri1 = ao2mo.restore(1, eri1, nmo)
        mo = cc._scf.mo_coeff
        Lpq = cc._scf._cderi
        ijslice = (0, nocc, nocc, nmo)
        Lov = _ao2mo.nr_e2(Lpq, mo, ijslice, aosym='s2', mosym='s1') 
        eris_ovov = Lov.T @ Lov
        self.ovov = eris_ovov.reshape(nocc, nvir, nocc, nvir)
        self.ovvo = self.ovov.transpose(0,1,3,2).copy()
        self.Lov = Lov


class drCCD(ccsd.CCSD):
    init_amps = init_amps
    update_amps = update_amps
    energy = energy
    ao2mo = ao2mo


if __name__ == '__main__':
	from pyscf import gto, dft, scf
	mol = gto.Mole()
	mol.verbose = 4
	mol.atom = 'H 0 0 0; F 0 0 1.1'
	mol.basis = 'ccpvdz'

	mol.build()
	mf = scf.RHF(mol).density_fit()
	mf.kernel()

	mycc = drCCD(mf)
	e_drccd, _, t2 = mycc.kernel()
	from pyscf.gw import rpa
	myrpa = rpa.RPA(mf)
	e_rpa = myrpa.kernel()
	print(f"Difference between drccd and drpa: {e_drccd-e_rpa} Ha." )

```

**Comment by tberkel** [comment]
Thanks a lot @changsuc for writing this. Since this question has come up multiple times, please feel free to turn your script into an example (to be placed into `examples/cc`) and submit a PR.
